#0: Update matmul sweep test with device data

skrsmanovicTT opened the pull request:
### Ticket
[Link to Github Issue ](https://github.com/tenstorrent/tt-llk/issues/370)

### Problem description

- Matmul benchmark results are only based on computation time measured by host that includes overhead introduced by dispatch as well.
- Size of chip grid to run test on is fixed to maximum available value, provide grid_size as optional argument to select different grid sizes.

### What's changed

- Added conftest.py fixture to support setting of chip grid size for test to run on. If not provided use maximum available grid size.
- Added device based utilization calculated using TRISC1 kernel length in number of cycles measured by default when profiler build is used. Kernel length is available in profiler log.
- If frequency throttling is applied during matmul OP utilization based on host timing can be missleading since it will use fixed frequency read by profiler to convert host measured time into number of cycles.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

skrsmanovicTT committed:
#0: Update matmul sweep test with device data

- Added device based utilization measure to the .csv report.
- Device utilization is based onTRISC1(Math) kernel length in number of
cycles.
- Profiler build always measures each RISC kernel length so this data
is awalible in profiler log.
- If throtling is enabled, host based utilziation can be missleading
since device frequency changes during test run and formula for
calcualting number of cycles measured by host is based on fixed
frequency reported in profiler log.
- Added conftest fixture for setting chip grid size for test to run on.

bbradelTT commented on tests/ttnn/unit_tests/benchmarks/test_benchmark.py:
```diff
@@ -476,9 +526,10 @@ def test_matmul_2d_host_perf_out_of_box(
-                "TFLOPs (avg)",
-                "Utilization (vs user grid)",
-                "Utilization (vs 8x8 full grid)",
+                f"Host based utilization[%] (vs {grid_size[0]}x{grid_size[1]} user selected grid)",
```
TFLOPs seems like it shouldn't be removed. And this looks like a header. I'm not sure if TFLOPs data is removed somewhere else.

skrsmanovicTT replied:
You are right, I accidentaly removed if only from header, addressing it with commit.

bbradelTT reviewed (commented):
Can you please provide examples of data provided before and after this change?

skrsmanovicTT committed:
#23713: Fix missing TFLOPs column in csv header

skrsmanovicTT commented:
@bbradelTT
I also updated description since runtime of sweep is fixed with this PR:
Size of profiler log file increases after every test iteration run (there is 106 iterations per one input configuration and there is in total 146+78=224 different input configurations) since profiler just append new data to existing log file. Parsing of log file is done after each new input configuration is tested(in order to get device frequency andTRISC1 kernel duration) and processing time is affected by log size. Since no old data is required, profiler log is cleared before every new input configuration is tested, so it includes only data relevant for input configuration under test. This reduces runtime to 20-30 minutes for complete sweep.

skrsmanovicTT commented:
Attaching same results when ppm is disabled, freq is set to fixed 1350MHz value
[main_ppm_disabled.csv](https://github.com/user-attachments/files/20798207/main_ppm_disabled.csv)
[pr_ppm_disabled.csv](https://github.com/user-attachments/files/20798208/pr_ppm_disabled.csv)
Need to double click difference in inference_time measured by host since it generates difference in host utilization as well

skrsmanovicTT committed:
#23713: Update matmul benchmark

- Added readme with more details about testing methodology and reported
metric results.
- Use fixed clock frequency value for calculating host baseed utilization.
- TODO replace hard coded frequency with a value read from device once
python support is implemented.
- Using frequency value read by profiler is not reliable since we can not
control moment at which profiler reads the value in case frequency
throtling is enabled.
- Synchronize device to ensure all warmup iterations are completed and
device is in clean state before measruing performance from host.

skrsmanovicTT commented:
@bbradelTT @yugaoTT @edwinleeTT I pushed proposed changes please leave some feedback.

I have:

1. Added readme docs and begging of file explaining methodology and metrics
2. Added fixed frequency value to be used for host based utilization calc. It is not ideal solution but at least I think it is worst case and can not produce results that are better than device utilization(which is non sense).  Unfortunately I am not aware of python api for reading clock value so we are kind of limited to using hard coded values in test itself.
3. Added sync point before we start measuring iterations to clean up CQ
4. Since there was no significant diff I did not force disabling profiler build for host utilization(for device utilization we must use profiler build)

If we want to measure something on kernel level like utilization of one matmul block, we would need to add zones inside matmul kernel .c code.

skrsmanovicTT committed:
#23713: Calculate device utilization only for profiler build

- Clean up comments and reademe
- Calculate device utilziation only when test is run with profiler build

skrsmanovicTT commented:
@bbradelTT 

I pushed hopefully final changes to the benchmark.

As I mentioned in https://github.com/tenstorrent/tt-metal/pull/23713#issuecomment-2988546287 I did not notice any obvious diff if we run with profiler build enabled or not. However I added this change to calculate host based utilization only if it is not profiler build. If it is profiler build it will calculate device utilization automatically since required data is in profiler log by default.

@bbradelTT do you know if all code owners are mandatory for PR to be merged?
